### PR TITLE
Construct SubMatrix/SubVector from numpy arrays without memcpy.

### PR DESCRIPTION
**Pull request by csukuangfj**
In addition, refactor kaldi pybind to be more modular.

It turns out that there is no way to construct a `Vector<float>` or `Matrix<float>` from a numpy
without data copy since `Vector<float>` and `Matrix<float>` take ownership of the memory and
will free it in their destructors.

### CODE REVIEW

File: src/pybind/Makefile [hunk 1]
```diff
@@ -33,7 +34,10 @@ ifeq ($(shell uname),Darwin)
   LDFLAGS += -undefined dynamic_lookup
 endif
 
-CCFILES = kaldi_pybind.cc
+CCFILES = kaldi_pybind.cc \
+					matrix/matrix_common_pybind.cc matrix/matrix_pybind.cc \
```
Comment (danpovey) [inline src/pybind/Makefile]: please use spaces instead of tabs here.

**Review by danpovey — COMMENTED** [review]
All looks fine but one small cosmetic issue.
How does one test for memory leakage with these types of interfaces?

**Comment by csukuangfj** [comment]
@danpovey 

that is a good point.

I found an approach mentioned in this post: https://stackoverflow.com/questions/3982036/how-can-i-use-valgrind-with-python-c-extensions

The output of `make valgrind` is

```
==9141== HEAP SUMMARY:
==9141==     in use at exit: 2,352,554 bytes in 4,496 blocks
==9141==   total heap usage: 66,348 allocs, 61,852 frees, 54,281,769 bytes allocated
==9141==
==9141== LEAK SUMMARY:
==9141==    definitely lost: 0 bytes in 0 blocks
==9141==    indirectly lost: 0 bytes in 0 blocks
==9141==      possibly lost: 144,881 bytes in 86 blocks
==9141==    still reachable: 2,207,673 bytes in 4,410 blocks
==9141==         suppressed: 0 bytes in 0 blocks
==9141== Rerun with --leak-check=full to see details of leaked memory
==9141==
==9141== Use --track-origins=yes to see where uninitialised values come from
==9141== For lists of detected and suppressed errors, rerun with: -s
==9141== ERROR SUMMARY: 2885 errors from 122 contexts (suppressed: 243 from 31)
```

I am not sure how many false positives there are in `possibly lost: 144,881 bytes in 86 blocks`

File: src/pybind/Makefile (hunk 1 above)
Comment (csukuangfj) [reply src/pybind/Makefile]: Done

**Comment by danpovey** [comment]
Try adding --leak-check=full --track-origins=yes to the valgrind command line in the Makefile, it might show what is happening.

**Comment by danpovey** [comment]
.. if there is a way to force garbage collection from Python, that might help reduce false positives.

**Comment by danpovey** [comment]
Anyway, merging.  Thanks a lot!

**Comment by danpovey** [comment]
It might actually be easiest to do manually memory debugging by putting a print statement in the destructor and making sure it's  called when you expect.

**Comment by csukuangfj** [comment]
I've tried to compile a debug version of CPython 3.5 from source following this page: https://pythonextensionpatterns.readthedocs.io/en/latest/debugging/valgrind.html

```sh
git clone https://github.com/python/cpython.git
cd cpython
git checkout 3.5

vim Objects/obmalloc.c --> go to line 1126  and enable the macro
#define Py_USING_MEMORY_DEBUGGER

mkdir valgrind
cd valgrind
../configure --with-pydebug --without-pymalloc --with-valgrind
make
```

The following command

```sh
 valgrind --tool=memcheck \
--leak-check=full --track-origins=yes \
--suppressions=/path/to/cpython/Misc/valgrind-python.supp \
/path/to/cpython/valgrind/python -E -tt -c "import gc; a=3; gc.collect()" -u gui,network
```
produces

```
==28941== LEAK SUMMARY:
==28941==    definitely lost: 0 bytes in 0 blocks
==28941==    indirectly lost: 0 bytes in 0 blocks
==28941==      possibly lost: 375,532 bytes in 1,124 blocks
==28941==    still reachable: 727,694 bytes in 5,960 blocks
==28941==         suppressed: 0 bytes in 0 blocks
==28941== Reachable blocks (those to which a pointer was found) are not shown.
==28941== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==28941==
==28941== For lists of detected and suppressed errors, rerun with: -s
==28941== ERROR SUMMARY: 287 errors from 287 contexts (suppressed: 0 from 0)
```

 `possibly lost: 375,532 bytes in 1,124 blocks` remains.

I think false positives in valgrind for testing python extensions cannot be avoided.

**Pull request merged**